File: src/main/java/SingleWindowScene.java
import org.lwjgl.LWJGLException;
import org.lwjgl.opengl.ContextAttribs;
import org.lwjgl.opengl.Display;
import org.lwjgl.opengl.DisplayMode;
import org.lwjgl.opengl.PixelFormat;

import static org.lwjgl.opengl.GL11.glViewport;

public abstract class SingleWindowScene implements Runnable {
  private final int width;
  private final int height;
  private final int openGlMajorVersion;
  private final int openGlMinorVersion;
  
  public SingleWindowScene(int width, int height, int openGlMajorVersion, int openGlMinorVersion) {
    this.width = width;
    this.height = height;
    
    this.openGlMajorVersion = openGlMajorVersion;
    this.openGlMinorVersion = openGlMinorVersion;
  }
  
  public void run() {
    try {
      configure();
      init();

      while (!Display.isCloseRequested()) {
        // Clear the screen and depth buffer
        display();
      }
      Display.destroy();
    } catch (LWJGLException e) {
      e.printStackTrace();
      System.exit(0);
    }
  }

  protected void configure() throws LWJGLException {
    PixelFormat pixelFormat = new PixelFormat();

    ContextAttribs contextAttribs = new ContextAttribs(openGlMajorVersion, openGlMinorVersion)
        .withForwardCompatible(true)
        .withProfileCore(true);

    Display.setDisplayMode(new DisplayMode(width, height));
    Display.setVSyncEnabled(true);
    Display.create(pixelFormat, contextAttribs);
    glViewport(0, 0, width, height);
  }

  protected abstract void init();
  protected abstract void display();
}

File: src/main/scala/SingleWindowScene.scala
package hello

import org.lwjgl.LWJGLException
import org.lwjgl.opengl.ContextAttribs
import org.lwjgl.opengl.Display
import org.lwjgl.opengl.DisplayMode
import org.lwjgl.opengl.PixelFormat
import org.lwjgl.opengl.GL11.glViewport


abstract class SingleWindowScene(width: Int, height: Int, openGlMajorVersion: Int, openGlMinorVersion: Int) extends Runnable {
  protected var lastFrameTime: Double = .0

  def run {
    try {
      configure()
      init()
      lastFrameTime = getTime
      while (!Display.isCloseRequested) {
        val deltaTime: Float = getDeltaTime
        display(deltaTime)
      }
      Display.destroy()
    }
    catch {
      case t: Throwable => {
        t.printStackTrace()
        if (Display.isCreated) Display.destroy()
        System.exit(-1)
      }
    }
    System.exit(0)
  }

  @throws(classOf[LWJGLException])
  protected def configure() {
    // http://stackoverflow.com/questions/18702467/enable-anti-aliasing-with-lwjgl
    // http://www.java-gaming.org/topics/lwjgl-antialiased-lines/30894/view.html
    val pixelFormat: PixelFormat = new PixelFormat(8, 8, 0, 8)
    val contextAttribs: ContextAttribs = new ContextAttribs(openGlMajorVersion,
                                                            openGlMinorVersion).withForwardCompatible(true).withProfileCore(true)
    Display.setDisplayMode(new DisplayMode(width, height))
    Display.setVSyncEnabled(true)
    Display.create(pixelFormat, contextAttribs)
    glViewport(0, 0, width, height)
  }

  protected def init(): Unit

  protected def display(delta: Float)

  def getTime: Double = {
    System.nanoTime / 1000000000.0
  }

  private def getDeltaTime: Float = {
    val time: Double = getTime
    val delta: Int = (time - lastFrameTime).toInt
    lastFrameTime = time
    delta
  }
}
